web/src/app/services/ServiceLabelFilterContainer.js

```diff
@@ -0,0 +1,61 @@
+import React from 'react'
+import p from 'prop-types'
+import Grid from '@material-ui/core/Grid'
+import Typography from '@material-ui/core/Typography'
+import { Filter as LabelFilterIcon } from 'mdi-material-ui'
+
+import { LabelKeySelect } from '../selection/LabelKeySelect'
+import { LabelValueSelect } from '../selection/LabelValueSelect'
+import FilterContainer from '../util/FilterContainer'
+
+export default function ServiceLabelFilterContainer(props) {
+  return (
+    <FilterContainer
+      icon={<LabelFilterIcon />}
+      title='Search by Labels'
+      iconButtonProps={{
+        'data-cy': 'services-filter-button',
+        color: 'default',
+        edge: 'end',
+        size: 'small',
+      }}
+      onReset={props.onReset}
+      anchorRef={props.anchorRef}
+    >
+      <Grid item xs={12}>
+        <Typography color='textSecondary'>
+          <i>Search by Label</i>
+        </Typography>
+      </Grid>
+      <Grid data-cy='label-key-container' item xs={12}>
+        <LabelKeySelect
+          name='label-key'
+          label='Select Label Key'
+          value={props.labelKey}
+          onChange={props.onKeyChange}
+        />
+      </Grid>
+      <Grid data-cy='label-value-container' item xs={12}>
+        <LabelValueSelect
+          name='label-value'
+          label='Select Label Value'
+          keyValue={props.labelKey}
+          value={props.labelValue}
+          onChange={props.onValueChange}
+          disabled={!props.labelKey}
+        />
+      </Grid>
+    </FilterContainer>
+  )
+}
+
+ServiceLabelFilterContainer.propTypes = {
+  labelKey: p.string,
```

Comment:
Instead of breaking out into key/value, can we use a single `onChange` and `value` that uses an object?
```js
value: p.shapeOf({key: p.string, value: p.string}),
onChange: p.func
```

It would clean up the interface a bit and be more consistent I think.

Reply:
I'm not really a fan of a `value.value` call, but we can certainly change it to `value.labelValue` and `value.labelKey` if we'd like. I don't think it really makes a difference with the interface/simplification, though, as we would still have to shape the object coming in.

Reply:
On second thought I looked into doing this, but I can't seem to find a clean way to have a single onChange handling different logic depending on key vs value, so I think it would be better to have them separated (as is). It's more clear that way to the reader in my opinion, too. The `onChange` logic can't be in the child (`<ServiceLabelFiltercontianer />`) because depending on the parent rendering it, that `onChange` will change for the `key` and `value` (setting in search params, vs setting in parent state).

Reply:
We should be able to do it in the same way we handle forms:

```js
function Container(props) {
  const {onChange, value} = props
  return (
    <KeyField value={value.labelKey} onChange={labelKey => onChange({...value, labelKey}) />
    <ValueField value={value.labelValue} onChange={labelValue => onChange({...value, labelValue}) />
  )
}
```

Reply:
As I see it, that would require some form of identifier as well so the parent knows which field to update accordingly in the URL, or storing the value in some object before storing it in the URL, using a useEffect for each type, neither of which I would prefer over my current implementation, which is quick, simple, and easy to read.

Reply:
The only field that should be stored in the URL is the search parameter, which will always be `key + '=' + value` and should be updated any time either value is changed. I think we're making this overly complicated, all that's needed (when combined) is:

```js
onChange={({ labelKey, labelValue }) =>
              setSearchParam(labelKey ? labelKey + '=' + labelValue : '')
            }
```

Here's the full change required to simplify this:
```patch
diff --git a/web/src/app/services/ServiceLabelFilterContainer.js b/web/src/app/services/ServiceLabelFilterContainer.js
index 73074c6..06792ff 100644
--- a/web/src/app/services/ServiceLabelFilterContainer.js
+++ b/web/src/app/services/ServiceLabelFilterContainer.js
@@ -9,6 +9,7 @@ import { LabelValueSelect } from '../selection/LabelValueSelect'
 import FilterContainer from '../util/FilterContainer'
 
 export default function ServiceLabelFilterContainer(props) {
+  const { labelKey, labelValue } = props.value
   return (
     <FilterContainer
       icon={<LabelFilterIcon />}
@@ -31,18 +32,20 @@ export default function ServiceLabelFilterContainer(props) {
         <LabelKeySelect
           name='label-key'
           label='Select Label Key'
-          value={props.labelKey}
-          onChange={props.onKeyChange}
+          value={labelKey}
+          onChange={labelKey => props.onChange({ ...props.value, labelKey })}
         />
       </Grid>
       <Grid data-cy='label-value-container' item xs={12}>
         <LabelValueSelect
           name='label-value'
           label='Select Label Value'
-          labelKey={props.labelKey}
-          value={props.labelValue}
-          onChange={props.onValueChange}
-          disabled={!props.labelKey}
+          labelKey={labelKey}
+          value={labelValue}
+          onChange={labelValue =>
+            props.onChange({ ...props.value, labelValue })
+          }
+          disabled={!labelKey}
         />
       </Grid>
     </FilterContainer>
@@ -50,10 +53,8 @@ export default function ServiceLabelFilterContainer(props) {
 }
 
 ServiceLabelFilterContainer.propTypes = {
-  labelKey: p.string,
-  labelValue: p.string,
-  onKeyChange: p.func.isRequired,
-  onValueChange: p.func.isRequired,
+  value: p.shape({ labelKey: p.string, labelValue: p.string }),
+  onChange: p.func,
   onReset: p.func,
 
   // optionally anchors the popover to a specified element's ref
diff --git a/web/src/app/services/ServiceRouter.js b/web/src/app/services/ServiceRouter.js
index cc4ff83..84ee5fc 100644
--- a/web/src/app/services/ServiceRouter.js
+++ b/web/src/app/services/ServiceRouter.js
@@ -38,13 +38,13 @@ export default function ServiceRouter() {
   const setSearchParam = value => dispatch(setURLParam('search', value))
 
   // grab key and value from the search param, if at all
-  let key = null
-  let value = null
+  let labelKey = ''
+  let labelValue = ''
   if (searchParam.includes('=')) {
     const searchSplit = searchParam.split(/(!=|=)/)
-    key = searchSplit[0]
+    labelKey = searchSplit[0]
     // the value can contain "=", so joining the rest of the match such that it doesn't get lost
-    value = searchSplit.slice(2).join('')
+    labelValue = searchSplit.slice(2).join('')
   }
 
   function renderList() {
@@ -62,10 +62,10 @@ export default function ServiceRouter() {
         createLabel='Service'
         searchAdornment={
           <ServiceLabelFilterContainer
-            labelKey={key}
-            labelValue={value}
-            onKeyChange={onKeyChange}
-            onValueChange={onValueChange}
+            value={{ labelKey, labelValue }}
+            onChange={({ labelKey, labelValue }) =>
+              setSearchParam(labelKey ? labelKey + '=' + labelValue : '')
+            }
             onReset={() => setSearchParam()}
           />
         }
@@ -73,25 +73,6 @@ export default function ServiceRouter() {
     )
   }
 
-  function onKeyChange(newKey) {
-    if (newKey) {
-      setSearchParam(newKey + '=')
-    } else {
-      setSearchParam() // clear search if clearing key
-    }
-  }
-
-  function onValueChange(newValue) {
-    // should be disabled if empty, but just in case :)
-    if (!key) return
-
-    if (newValue) {
-      setSearchParam(key + '=' + newValue)
-    } else {
-      setSearchParam(key + '=')
-    }
-  }
-
   function renderDetails({ match }) {
     return <ServiceDetails serviceID={match.params.serviceID} />
   }
```


----

We should always be using `value` and `onChange` for value passing and change callbacks where we can so they work the same as React, MUI, react-select, etc... Sometimes it's going to require a bit of extra work and/or planning but it keeps the component interfaces small, predictable and consistent.

Reply:
Feel free to commit.. I was happy with my implementation

Reply:
I don't think we should force ourselves into an internal API when not needed, though, so I will have to disagree with you there.